ctapipe_io_lst/calibration.py

```diff
@@ -0,0 +1,778 @@
+            container = event.r1.tel[tel_id]
+        else:
+            # Modify R0 container. This is to create pedestal files.
+            container = event.r0.tel[tel_id]
+
+        waveform = container.waveform.copy()
+
+        # We have 2 functions: one for data from 2018/10/10 to 2019/11/04 and
+        # one for data from 2019/11/05 (from Run 1574) after update firmware.
+        # The old readout (before 2019/11/05) is shifted by 1 cell.
+        run_id = event.lst.tel[tel_id].svc.configuration_id
+        if run_id > LAST_RUN_WITH_OLD_FIRMWARE:
+            time_lapse_corr = do_time_lapse_corr
+        else:
+            time_lapse_corr = do_time_lapse_corr_data_from_20181010_to_20191104
+
+        time_lapse_corr(
+            waveform,
+            lst.evt.local_clock_counter,
+            self.first_cap[tel_id],
+            self.last_readout_time[tel_id],
+            lst.svc.pixel_ids,
+        )
+
+        container.waveform = waveform
+
+    def interpolate_spikes(self, event, tel_id):
+        """
+        Interpolates spike A & B.
+        Fill the R1 container.
+        Parameters
+        ----------
+        event : `ctapipe` event-container
+        tel_id : id of the telescope
+        """
+        run_id = event.lst.tel[tel_id].svc.configuration_id
+
+        # Interpolate spikes should be done after pedestal subtraction and time lapse correction.
+        if isinstance(event.r1.tel[tel_id].waveform, np.ndarray):
+            waveform = event.r1.tel[tel_id].waveform.copy()
+
+            # We have 2 functions: one for data from 2018/10/10 to 2019/11/04 and
+            # one for data from 2019/11/05 (from Run 1574) after update firmware.
+            # The old readout (before 2019/11/05) is shifted by 1 cell.
+            if run_id > LAST_RUN_WITH_OLD_FIRMWARE:
+                interpolate_pseudo_pulses = self.interpolate_pseudo_pulses
+            else:
+                interpolate_pseudo_pulses = self.interpolate_pseudo_pulses_data_from_20181010_to_20191104
+
+            interpolate_pseudo_pulses(
+                waveform,
+                self.first_cap[tel_id],
+                self.first_cap_old[tel_id],
+            )
+            event.r1.tel[tel_id].waveform = waveform
+
+    @staticmethod
+    @njit(cache=True)
+    def interpolate_pseudo_pulses(waveform, fc, fc_old):
+        """
+        Interpolate Spike type A. Modifies waveform in place
+
+        Parameters
+        ----------
+        waveform : ndarray
+            Waveform stored in a numpy array of shape
+            (N_GAINS, N_PIXELS, N_SAMPLES).
+        fc : ndarray
+            Value of first capacitor stored in a numpy array of shape
+            (N_GAINS, N_PIXELS).
+        fc_old : ndarray
+            Value of first capacitor from previous event
+            stored in a numpy array of shape
+            (N_GAINS, N_PIXELS).
+        """
+        LAST_IN_FIRST_HALF = N_CAPACITORS_CHANNEL // 2 - 1
+
+        for gain in range(N_GAINS):
```

Comment:
Wouldn't `prange` from `numba` speed up the code?

Reply:
In my tests it didn't, it was even slower. The arrays are not very large, so the overhead of spawning a new process is higher.

Also: this code will mostly run on large clusters, where we parallelize by running many jobs that use a single core, not by running one job that uses many cores.